internal/pkg/gateway/commit/blockparser.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2021 IBM All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package commit
+
+import (
+	"errors"
+
+	"github.com/golang/protobuf/proto"
+	commonProto "github.com/hyperledger/fabric-protos-go/common"
+	peerProto "github.com/hyperledger/fabric-protos-go/peer"
+	ghErrors "github.com/pkg/errors"
+)
+
+type blockParser struct {
+	Block *commonProto.Block
+}
+
+func (parser *blockParser) TransactionValidationCodes() (map[string]peerProto.TxValidationCode, error) {
+	results := make(map[string]peerProto.TxValidationCode)
+
+	envelopes := parser.Block.Data.Data
+	for i, envelopeBytes := range envelopes {
+		channelHeader, err := unmarshalChannelHeader(envelopeBytes)
+		if err != nil {
+			return nil, ghErrors.WithMessagef(err, "failed to unmarshal channel header from envelope at index %v in block number %v",
+				i, parser.Block.Header.Number)
+		}
+
+		validationCode := parser.validationCode(i)
+		if validationCode == peerProto.TxValidationCode_BAD_PROPOSAL_TXID {
```

Comment:
I understand this code since we discussed it, but perhaps a small comment explaining that why we single out this particular status type.  

```
// Ignore txids which could have been forged by other parties,
// otherwise accept the first failure since it originated from the gateway client.
```